File: src/test/java/edu/kit/iti/formal/stvs/logic/specification/smtlib/SExpressionTest.java
package edu.kit.iti.formal.stvs.logic.specification.smtlib;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;

/**
 * Created by csicar on 08.02.17.
 */
@RunWith(Parameterized.class)
public class SExpressionTest {
  public SExpression expression;


  public SExpressionTest(SExpression expression) {
    this.expression = expression;
  }

  @Parameterized.Parameters
  public static Collection<Object[]> instancesToTest() {
    SList list = new SList("a", new SAtom("b"), new SAtom("c"), new SList("n1", "n2"));
    return Arrays.asList(
        new Object[]{list},
        new Object[]{new SAtom("asdasd")},
        new Object[]{new SAtom("1")}
    );
  }

  @Test
  public void fromString() throws Exception {
    assertEquals(
        new SList("a", new SAtom("b"), new SAtom("c"),
            new SList("nested1", "2", "3"),
            new SList("4")),
        SExpression.fromText("(a b c (nested1 2 3 ) (4))"));
  }

  @Test(expected = IllegalArgumentException.class)
  public void fromStringWrongInput() {
    SExpression.fromText(")asdasd");
  }

  @Test
  public void fromSexp() throws Exception {
    assertEquals(expression, SExpression.fromSexp(expression.toSexpr()));
  }

  @Test
  public void isAtom() throws Exception {
    assertEquals(expression.toSexpr().isAtomic(), expression.isAtom());
  }

  @Test
  public void toText() throws Exception {
    assertEquals(expression, SExpression.fromText(expression.toText()));
  }

  @Test
  public void testEquals() {
    assertNotEquals(expression, this);
    assertFalse(expression.equals(null));
  }
}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/logic/specification/smtlib/SExpressionTest.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.logic.specification.smtlib

import com.google.common.truth.Truth.assertThat
import edu.kit.iti.formal.smt.SExpr
import edu.kit.iti.formal.smt.SExprFacade.parseExpr
import edu.kit.iti.formal.smt.SExprFacade.sexpr
import edu.kit.iti.formal.smt.SList
import edu.kit.iti.formal.smt.SNumber
import edu.kit.iti.formal.smt.SSymbol
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import kotlin.test.assertFailsWith

internal fun fromText(s: String) = parseExpr(s)

/**
 * Created by csicar on 08.02.17.
 */
class SExpressionTest {
    @Test
    fun fromString() {
        assertThat(SSymbol("a")).isEqualTo(SSymbol("a"))
        assertThat(SList(SSymbol("a"))).isEqualTo(SList(SSymbol("a")))
        assertThat(fromText("a")).isEqualTo(SSymbol("a"))
        assertThat(fromText("()")).isEqualTo(SList())
        assertThat(fromText("(a)")).isEqualTo(SList(SSymbol("a")))

        assertThat(fromText("(nested1 2 3)"))
            .isEqualTo(sexpr("nested1", 2, 3))

        assertThat(
            fromText("(a b c (nested1 2 3 ) (4))"),
        ).isEqualTo(
            SList(
                "a",
                SSymbol("b"),
                SSymbol("c"),
                sexpr("nested1", 2, 3),
                SList(SNumber(4)),
            ),
        )
    }

    @Test
    fun fromStringWrongInput() {
        assertFailsWith<IllegalArgumentException> {
            fromText(")asdasd")
        }
    }

    @ParameterizedTest
    @MethodSource("instancesToTest")
    fun toText(expression: SExpr) {
        assertThat(fromText(expression.toString())).isEqualTo(expression)
    }

    @ParameterizedTest
    @MethodSource("instancesToTest")
    fun testEquals(expression: SExpr) {
        assertNotEquals(expression, this)
        assertEquals(expression, expression)
        assertFalse(false)
    }

    companion object {
        @JvmStatic
        fun instancesToTest(): List<Arguments> {
            val list = SList("a", SSymbol("b"), SSymbol("c"), sexpr("n1", "n2"))
            return listOf(
                Arguments.of(list),
                Arguments.of(SSymbol("asdasd")),
                Arguments.of(SNumber(1)),
            )
        }
    }
}